src/nvmet.c

```diff
@@ -833,209 +909,230 @@ int nvmet_revalidate_size(struct RD_PROFILE *rd_prof, RC_PROFILE *rc_prof, char
+	int rc = INVALID_VALUE, n, allowed_host_number, rv = INVALID_VALUE, port_number;
 	FILE *fp;
-	char hostname[0x40] = {0x0}, path[NAMELEN] = {0x0};
+	char hostname[0x40] = {0x0}, path[NAMELEN] = {0x0}, nqn[NAMELEN] = {0x0};
+	mode_t mode = (S_IRUSR | S_IWUSR | S_IXUSR | S_IRGRP | S_IXGRP | S_IROTH | S_IXOTH);
 	struct dirent **list;
 	char *msg;
-	char json_message[NAMELEN] = {0};
-	char json_message_temp[NAMELEN] = {0};
+	char target_string[NAMELEN] = {0};
+	char target_string_final[NAMELEN] = {0};
+	char regexp[NAMELEN] = {0};
+	struct NVMET_PROFILE *profile;
+	struct NVMET_PROFILE *orig_profile;
+	struct NVMET_PORTS *ports;
+	struct NVMET_PORTS *orig_ports;
 
-	gethostname(hostname, sizeof(hostname));
-	sprintf(path, "%s/%s%s-%s", SYS_NVMET_TGT, NQN_HDR_STR, hostname, device);
-	if (access(path, F_OK) != SUCCESS) {
-		sprintf(path, "%s%s-%s", NQN_HDR_STR, hostname, device);
-		msg = "Error. NQN export: %s does not exist.";
-		print_error(msg, return_message, path);
+	if (gethostname(hostname, sizeof(hostname)) < 0) {
+		msg = "Error. Unable to obtain hostname. %s: gethostname: %s";
+		print_error(msg, return_message, __func__, strerror(errno));
 		return rc;
 	}
 
-	/* Check if namespace 1 exists. That is the only namespace we define. */
-	sprintf(path, "%s/%s%s-%s/namespaces/1", SYS_NVMET_TGT, NQN_HDR_STR, hostname, device);
-	if (access(path, F_OK) != SUCCESS) {
-		msg = "%s: A RapidDisk defined namespace does not exist.";
-		print_error(msg, return_message, __func__);
-		return rc;
-	}
+	sprintf(nqn, "%s%s-%s", NQN_HDR_STR, hostname, device);
 
-	/* Make sure that no other namespaces exist. We do not create anything higher than 1. */
-	sprintf(path, "%s/%s%s-%s/namespaces/", SYS_NVMET_TGT, NQN_HDR_STR, hostname, device);
-	if ((err = scandir(path, &list, scandir_filter_no_dot, NULL)) < 0) {
-		msg = "Error. Unable to access %s. %s: scandir: %s";
-		print_error(msg, return_message, path, __func__, strerror(errno));
-		return INVALID_VALUE;
-	}
+	if (port != INVALID_VALUE) {
+		ports = nvmet_scan_ports(return_message, &rv);
+		if (rv == INVALID_VALUE) {
+			return rc;
+		}
+		orig_ports = ports;
+		while (ports != NULL) {
+			if (port == ports->port) {
+				rc = 1;
+				if (strcmp(nqn, ports->nqn) == 0) {
+					rc = SUCCESS;
+					break;
+				}
+			}
+			ports = ports->next;
+		}
+		if (rc == INVALID_VALUE) {
+			print_error(ERR_PORT_NOEXIST, return_message, port);
+			free_nvmet_linked_lists(orig_ports, NULL);
+			return rc;
+		} else if (rc == 1) {
+			print_error("NQN %s does not exists!", return_message, nqn);
+			free_nvmet_linked_lists(orig_ports, NULL);
+			return INVALID_VALUE;
+		}
 
-	list = clean_scandir(list, err);
+		/* Reset rc to INVALID_VALUE after port check */
+		rc = INVALID_VALUE;
 
-	if (err > 1) {
-		msg = "An invalid number of namespaces not created by RapidDisk exist.";
-		print_error("%s", return_message, msg);
-		return rc;
+		free_nvmet_linked_lists(orig_ports, NULL);
+	} else {
```

Comment:
@matteotenca: I feel like this entire else statement could easily be replace with a:
sprintf(path, "%s/%s", SYS_NVMET_TGT, nqn)
if (access(path, F_OK) != SUCCESS) {
    print_error("NQN %s does not exists!", return_message, nqn);
    return rc;
}

What are your thoughts on this?

Reply:
Well, I believe I wrote this code to mimic the `nvmet_export_volume` function: I believe you created a nice set of linked lists which can be populated with data regarding the ongoing situation about mappings/ramdisks/nvmet, so I went this way, utilizing them. For sure, at low level, it's always a matter of which dirs/files were created/are present/contain some values in kernel's filesystem. But I thought it would be nice to move up a level, and delegate the low level checks to functions that populate the linked lists.

Tell me if I missed the point!